src/core/metadata_manager.py

```diff
@@ -84,73 +65,48 @@ def _extractors(cls) -> list[Type[MetadataBase]]:
-                return False
-        return True
-
-    async def _extract_meta_data(
-        self,
-        site: WebsiteData,
-        allow_list: Optional[list[str]],
-    ) -> dict:
-        data = {}
-        tasks = []
-
-        async def run_extractor_with_key(key: str, extractor: MetadataBase, site: WebsiteData) -> tuple[str, dict]:
-            """Call the extractor and pack its results into a tuple containing also the key"""
-            try:
-                duration, values, stars, explanation = await extractor.start(site=site)
-                return key, {
-                    VALUES: values,
-                    TIME_REQUIRED: duration,
-                    STAR_CASE: stars,
-                    EXPLANATION: explanation,
-                }
-            except Exception as e:
-                self._logger.exception(f"Failed to extract from {key}: {e}")
-                return key, {"exception": str(e)}
+        async def create_extractor(Extractor: Type[MetadataBase]) -> MetadataBase:  # noqa
+            instance = Extractor()
+            await instance.setup()
+            return instance
 
-        for extractor in self.metadata_extractors:
-            if allow_list is None or extractor.key in allow_list:
-                tasks.append(run_extractor_with_key(key=extractor.key, extractor=extractor, site=site))
-
-        results: list[tuple[str, dict]] = await asyncio.gather(*tasks)
-
-        # combine the cached results with the newly extracted data into the target dictionary structure
-        return {
-            **data,
-            **{k: d for k, d in results},
-        }
+        logging.info("Initializing extractors")
+        extractors: tuple[MetadataBase, ...] = await asyncio.gather(
+            *[create_extractor(extractor) for extractor in types]
+        )
+        logging.info("Creating MetadataManager")
+        return MetadataManager(extractors=list(extractors))
 
-    async def start(self, message: Message) -> dict:
-        self._logger.debug(f"Start metadata_manager at {time.perf_counter() - global_start} since start")
+    async def extract(self, message: Input) -> Output:
+        self.logger.debug("Calling extractors from manager")
 
         # this will eventually load the content dynamically if not provided in the message
         # hence it may fail with various different exceptions (ConnectionError, ...)
         # those exceptions should be handled in the caller of this function.
-        site = WebsiteData.from_message(
-            message=message,
-            logger=self._logger,
+        site = await WebsiteData.from_input(
+            input=message,
+            logger=self.logger,
             tld_extractor=self.tld_extractor,
         )
 
-        self._logger.debug(f"WebsiteManager loaded at {time.perf_counter() - global_start} since start")
+        self.logger.debug("Build website data object.")
+
+        async def run_extractor(extractor: MetadataBase) -> Union[MetadataTags, Error]:
```

Comment:
Its only used in this function -> should not be part of the public api. AFAIK there are no async lambdas, hence the explicit function declaration.

The pros of having it nested: no one else can see it becomes immediately clear if it is no longer required. I like it because it prevents piling up unused code and growing coupling between files.

Reply:
I agree with your pro points, the first one could be solved with __ as well, due to name wrangling.
Maybe I just need to get used to it :/